xds/internal/resolver/xds_resolver.go

```diff
@@ -137,67 +169,83 @@ func (b *xdsResolverBuilder) Build(target resolver.Target, cc resolver.ClientCon
+	logger *grpclog.PrefixLogger
 	// The underlying xdsClient which performs all xDS requests and responses.
 	xdsClient      xdsclient.XDSClient
 	xdsClientClose func()
-	// A channel for the watch API callback to write service updates on to. The
-	// updates are read by the run goroutine and passed on to the ClientConn.
-	updateCh chan suWithError
-	// cancelWatch is the function to cancel the watcher.
-	cancelWatch func()
-
-	// activeClusters is a map from cluster name to a ref count.  Only read or
-	// written during a service update (synchronous).
+	// A random number which uniquely identifies the channel which owns this
+	// resolver.
+	channelID uint64
+
+	// All methods on the xdsResolver type except for the ones invoked by gRPC,
+	// i.e ResolveNow() and Close(), are guaranteed to execute in the context of
+	// this serializer's callback. And since the serializer guarantees mutual
+	// exclusion among these callbacks, we can get by without any mutexes to
+	// access all of the below defined state. The only exception is Close(),
+	// which does access some of this shared state, but it does so after
+	// cancelling the context passed to the serializer.
+	serializer       *grpcsync.CallbackSerializer
+	serializerCancel context.CancelFunc
+
+	// Listener watcher related state.
+	ldsResourceName     string
+	listenerWatcher     *listenerWatcher
+	listenerUpdateRecvd bool
+	currentListener     xdsresource.ListenerUpdate
+
+	// RouteConfiguration watcher related state.
+	rdsResourceName        string
+	routeConfigWatcher     *routeConfigWatcher
+	routeConfigUpdateRecvd bool
+	currentRouteConfig     xdsresource.RouteConfigUpdate
+	currentVirtualHost     *xdsresource.VirtualHost // Matched virtual host for quick access.
+
+	// activeClusters is a map from cluster name to information about the
+	// cluster that includes a ref count and load balancing configuration.
 	activeClusters map[string]*clusterInfo
 
 	curConfigSelector *configSelector
+}
 
-	// A random number which uniquely identifies the channel which owns this
-	// resolver.
-	channelID uint64
+// ResolveNow is a no-op at this point.
+func (*xdsResolver) ResolveNow(o resolver.ResolveNowOptions) {}
+
+func (r *xdsResolver) Close() {
+	// Cancel the context passed to the serializer and wait for any scheduled
+	// callbacks to complete. Canceling the context ensures that no new
+	// callbacks will be scheduled.
+	r.serializerCancel()
+	<-r.serializer.Done()
+
+	// Note that Close needs to check for nils even if some of them are always
+	// set in the constructor. This is because the constructor defers Close() in
+	// error cases, and the fields might not be set when the error happens.
+
+	if r.listenerWatcher != nil {
+		r.listenerWatcher.stop()
+	}
+	if r.routeConfigWatcher != nil {
+		r.routeConfigWatcher.stop()
+	}
+	if r.xdsClientClose != nil {
+		r.xdsClientClose()
+	}
+	r.logger.Infof("Shutdown")
 }
 
 // sendNewServiceConfig prunes active clusters, generates a new service config
```

Comment:
Please mention that this can only be executing inside a serializer callback. It's pretty deep within the call chain, but I don't want to break this assumption in the future by calling this outside of functions ran in the callback serializer.